File: java/src/main/java/toolgood/algorithm/internals/functions/string/Function_ASC.java
package toolgood.algorithm.internals.functions.string;

import java.util.List;

import toolgood.algorithm.AlgorithmEngine;
import toolgood.algorithm.Operand;
import toolgood.algorithm.enums.OperandType;
import toolgood.algorithm.internals.ParameterType;
import toolgood.algorithm.internals.functions.FunctionBase;
import toolgood.algorithm.internals.functions.Function_1;
import toolgood.algorithm.internals.functions.NoneEngine;

public final class Function_ASC extends Function_1 {
    public Function_ASC(FunctionBase func1) {
        super(func1);
    }

    @Override
    public String Name() {
        return "Asc";
    }

    @Override
    public Operand Evaluate(AlgorithmEngine engine, java.util.function.BiFunction<AlgorithmEngine, String, Operand> tempParameter) throws Exception {
        Operand args1 = GetText_1(engine, tempParameter);
        if (args1.IsErrorOrNone()) { return args1; }
        return Operand.Create(F_base_ToDBC(args1.TextValue()));
    }

    private static String F_base_ToDBC(String input) {
        boolean needModify = false;
        for (int i = 0; i < input.length(); i++) {
            char c = input.charAt(i);
            if (c == 12288 || (c > 65280 && c < 65375)) {
                needModify = true;
                break;
            }
        }
        if (!needModify) {
            return input;
        }
        char[] chars = input.toCharArray();
        for (int i = 0; i < chars.length; i++) {
            char c = chars[i];
            if (c == 12288) {
                chars[i] = (char) 32;
            } else if (c > 65280 && c < 65375) {
                chars[i] = (char) (c - 65248);
            }
        }
        return new String(chars);
    }

    @Override
    public OperandType GetResultType() {
        return OperandType.TEXT;
    }

    @Override
    public void GetParameterTypes(NoneEngine noneEngine, List<ParameterType> result, OperandType operandType, String op, String val) {
        func1.GetParameterTypes(noneEngine, result, OperandType.TEXT);
    }
}

File: csharp/ToolGood.Algorithm/Internals/Functions/String/Function_ASC.cs
using System;
using System.Text;

namespace ToolGood.Algorithm.Internals.Functions.String
{
	internal class Function_ASC : Function_1
	{
		public Function_ASC(FunctionBase func1) : base(func1)
		{
		}

		public override string Name => "Asc";

		public override Operand Evaluate(AlgorithmEngine work, Func<AlgorithmEngine, string, Operand> tempParameter)
		{
			var args1 = GetText_1(work, tempParameter);
			if (args1.IsError) { return args1; }
			return Operand.Create(F_base_ToDBC(args1.TextValue));
		}

		private static string F_base_ToDBC(string input)
		{
			bool needModify = false;
			for (int i = 0; i < input.Length; i++) {
				var c = input[i];
				if (c == 12288 || (c > 65280 && c < 65375)) {
					needModify = true;
					break;
				}
			}
			if (!needModify) {
				return input;
			}
			char[] chars = input.ToCharArray();
			Span<char> span = chars;
			for (int i = 0; i < span.Length; i++) {
				var c = span[i];
				if (c == 12288) {
					span[i] = (char)32;
				} else if (c > 65280 && c < 65375) {
					span[i] = (char)(c - 65248);
				}
			}
			return new string(chars);
		}

	}

}
